bolt-modules/boltlib/lib/puppet/functions/run_plan.rb

```diff
@@ -85,6 +85,22 @@ def run_plan(scope, plan_name, named_args = {})
       executor.report_yaml_plan(closure.model.body)
     end
 
+    # If a TargetSpec parameter is passed, ensure it is in inventory
+    inventory = Puppet.lookup(:bolt_inventory)
+
+    if inventory.version > 1
+      param_types = closure.parameters.each_with_object({}) do |param, param_acc|
+        param_acc[param.name] = extract_parameter_types(param.type_expr).flatten
+      end
+      params.each do |param, value|
+        # Note the safe lookup operator is needed to handle case where a parameter is passed to a
+        # plan that the plan is not expecting
+        if param_types[param]&.include?('TargetSpec') || param_types[param]&.include?('Boltlib::TargetSpec')
+          inventory.ensure_in_all_group(value)
```

Comment:
Is this different than `inventory.get_targets(value)` other than doing a bit less computation?

Reply:
The main difference is that get_targets will add a target to inventory, but it will not add it to the all group. Effectively an inventory's `@target` hash will have the new target, but that does not necessarily add the target to the all group. My understanding was that we only add a target to all with `Target.new`, `get_target` or if a plan has a `TargeSpec` parameter. This is based off the "riddle" i posted https://puppet.slack.com/archives/CF466J51Q/p1566581349473200

Reply:
```
cas donoghue 10:29 AM
With an empty inventory:
version: 2
a plan like:
plan test(TargetSpec $nodes){
  notice(get_targets('all')
}
And an invocation:
bolt plan run test -t localhost
Will localhost be in the targets array?
Nick Lewis:boltlogo: 10:31 AM
are you the sphinx??
:rolling_on_the_floor_laughing:
1

I’m going with no, but I’m sure not confident about it!
David Kramer:boltlogo: 10:32 AM
BoltTeasers™
Nick Lewis:boltlogo: 10:33 AM
I would potentially think it will not be in the inventory, but if you call $nodes.get_targets first then it will be in the inventory for get_targets('all')
But I also think that is not the actual behavior
cas donoghue 10:37 AM
I think this relates to Target.new in a plan. It is pretty easy to add a new target in inventory with get_target, but Target.new does not yet have an inventory attribute. It would be pretty easy to add $nodes to inventory but if they are not in inventory, why would Target.new be?
Nick Lewis:boltlogo: 10:40 AM
I distinctly remember having discussions about this, but I don’t remember the conclusion.
I think calling Target.new should add the target to the inventory.
Calling it from a plan, specifically.
```
I guess I just read the frist answer ("i'm going with no") and rolled with it. Reading it again it seems like we (at least in that thread) did not come to a conclusion other than me implementing it the way I did.

Reply:
Though this example would be in inventory. Sorry wrong link. I don't remember when we talked about get_targets. Maybe I made it up.

Reply:
I see. I think the context around that question _at the time_ was whether targets provided on the CLI are automatically added to the inventory at all (irrespective of the "all" group). Since we later (IIRC) decided that arguments of type `TargetSpec` should be automatically added to the inventory, the behavior of that example changes.

Reply:
Updated